src/zkp/ecdsa_adaptor.rs

```diff
@@ -0,0 +1,317 @@
+        let mut res = [0; ECDSA_ADAPTOR_SIGNATURE_LENGTH];
+        match from_hex(s, &mut res) {
+            Ok(ECDSA_ADAPTOR_SIGNATURE_LENGTH) => {
+                EcdsaAdaptorSignature::from_slice(&res[0..ECDSA_ADAPTOR_SIGNATURE_LENGTH])
+            }
+            _ => Err(Error::InvalidEcdsaAdaptorSignature),
+        }
+    }
+}
+
+impl CPtr for EcdsaAdaptorSignature {
+    type Target = ffi::EcdsaAdaptorSignature;
+    fn as_c_ptr(&self) -> *const Self::Target {
+        self.as_ptr()
+    }
+
+    fn as_mut_c_ptr(&mut self) -> *mut Self::Target {
+        self.as_mut_ptr()
+    }
+}
+
+impl EcdsaAdaptorSignature {
+    /// Creates an AdaptorSignature directly from a slice
+    #[inline]
+    pub fn from_slice(data: &[u8]) -> Result<EcdsaAdaptorSignature, Error> {
+        match data.len() {
+            ECDSA_ADAPTOR_SIGNATURE_LENGTH => {
+                let mut ret = [0; ECDSA_ADAPTOR_SIGNATURE_LENGTH];
+                ret[..].copy_from_slice(data);
+                Ok(EcdsaAdaptorSignature(ffi::EcdsaAdaptorSignature::from(ret)))
+            }
+            _ => Err(Error::InvalidEcdsaAdaptorSignature),
+        }
+    }
+
+    /// Obtains a raw const pointer suitable for use with FFI functions
+    #[inline]
+    pub fn as_ptr(&self) -> *const ffi::EcdsaAdaptorSignature {
+        &self.0
+    }
+
+    /// Obtains a raw mutable pointer suitable for use with FFI functions
+    #[inline]
+    pub fn as_mut_ptr(&mut self) -> *mut ffi::EcdsaAdaptorSignature {
+        &mut self.0
+    }
+}
+
+impl EcdsaAdaptorSignature {
+    /// Creates an adaptor signature along with a proof to verify the adaptor signature.
+    pub fn encrypt<C: Signing>(
+        secp: &Secp256k1<C>,
+        msg: &Message,
+        sk: &SecretKey,
+        enckey: &PublicKey,
+    ) -> Result<EcdsaAdaptorSignature, Error> {
+        let mut adaptor_sig = ffi::EcdsaAdaptorSignature::new();
+
+        let ret = unsafe {
+            ffi::secp256k1_ecdsa_adaptor_encrypt(
+                *secp.ctx(),
+                &mut adaptor_sig,
+                sk.as_c_ptr(),
+                enckey.as_c_ptr(),
+                msg.as_c_ptr(),
+                None,
+                ptr::null_mut(),
+            )
+        };
+
+        if ret != 1 {
+            return Err(Error::InvalidEcdsaAdaptorSignature);
+        }
+
+        Ok(EcdsaAdaptorSignature(adaptor_sig))
+    }
+
+    /// Creates an ECDSA signature from an adaptor signature and an adaptor secret.
+    pub fn decrypt(&self, decryption_key: &SecretKey) -> Result<Signature, Error> {
+        unsafe {
+            let mut signature = ffi::Signature::new();
+            let ret = ffi::secp256k1_ecdsa_adaptor_decrypt(
+                ffi::secp256k1_context_no_precomp,
+                &mut signature,
+                decryption_key.as_c_ptr(),
+                self.as_c_ptr(),
+            );
+
+            if ret != 1 {
+                return Err(Error::InvalidEcdsaAdaptorSignature);
```

Comment:
Looking at our other error variants we are actually quite detailed there. For example we have `CannotProveSurjection` and `InvalidSurjectionProof`.

I think we should do the same thing here and have variants for:

- `CannotDecryptAdaptorSignature`
- `CannotRecoverAdaptorSecret`
- `CannotVerifyAdaptorSignature` (this is different from invalid I'd say because one says that we could parse it and the other one says that the combination of signature, adaptor signature and keys doesn't work)

Reply:
Fair enough, I added the suggested errors. Maybe check the documentation for them because I'm not 100% confident it's good.

Reply:
Documentation is good, thank you!